src/main/java/io/fabric8/vertx/maven/plugin/utils/IncrementalBuilder.java

```diff
@@ -17,108 +17,131 @@
+        inclDirs.forEach(this::buildObserver);
 
+    }
 
-        Consumer<Path> fnPathRegistration = path -> {
+    @Override
+    public void run() {
+        try {
+            this.monitor.start();
+        } catch (Exception e) {
+            logger.error("Unable to start Incremental Builder", e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (this.monitor != null) {
             try {
-                register(path);
-            } catch (IOException e) {
-                throw new UncheckedIOException(e);
+                this.monitor.stop();
+            } catch (Exception e) {
+                //ignore
             }
-        };
-
-        inclDirs.forEach(fnPathRegistration);
+        }
     }
 
-    @SuppressWarnings("unchecked")
-    static <T> WatchEvent<T> cast(WatchEvent<?> event) {
-        return (WatchEvent<T>) event;
-    }
+    /**
+     * @param path
+     */
+    protected synchronized void buildObserver(Path path) {
 
-    @Override
-    public void run() {
+        logger.info("Observing path:" + path.toString());
 
-        logger.info("Incremental builder watching paths: " + watchPathKeys.values());
+        FileAlterationObserver observer = new FileAlterationObserver(path.toFile());
 
-        while (true) {
+        observer.addListener(this);
 
-            WatchKey key;
+        observers.put(path, observer);
 
-            try {
-                key = fileWatcherService.take();
-            } catch (InterruptedException e) {
-                return;
-            }
+        this.monitor.addObserver(observer);
+    }
 
-            watchPathKeys.get(key);
+    /**
+     *
+     */
+    protected synchronized void syncMointor() {
+        observers.forEach((path, observer) -> {
+            this.monitor.getObservers().forEach(observer2 -> {
+                Path path1 = Paths.get(observer2.getDirectory().toString());
+                if (!observers.containsKey(path1)) {
+                    this.monitor.removeObserver(observer2);
+                }
+            });
+        });
+    }
 
-            WatchEvent.Kind<?> kind = null;
-            for (WatchEvent<?> watchEvent : key.pollEvents()) {
-                kind = watchEvent.kind();
 
-                if (StandardWatchEventKinds.ENTRY_MODIFY == kind ||
-                        StandardWatchEventKinds.ENTRY_CREATE == kind ||
-                        StandardWatchEventKinds.ENTRY_DELETE == kind) {
+    @Override
+    public void onDirectoryCreate(File directory) {
+        buildObserver(Paths.get(directory.toString()));
+        syncMointor();
+    }
 
-                    WatchEvent<Path> pathEvent = cast(watchEvent);
-                    Path path = pathEvent.context();
 
-                    this.callback.call(kind, path);
+    @Override
+    public void onDirectoryDelete(File directory) {
+        observers.remove(Paths.get(directory.toString()));
```

Comment:
I don't think so  we need to deregister the file observers as the observers are registered at directory level